### ISSUE TITLE
Add option to automatically convert TermsFacet to NestedFacet

### ISSUE THREAD

**Issue by melissachang** [elastic/elasticsearch-dsl-py#1034]
For my use case, for nested fields I always want NestedFacet instead of TermsFacet. (I would think this is true for many use cases... maybe at least 50%.)

It takes a non-trivial amount of code to [convert all relevant TermsFacet's to NestedFacets's](https://github.com/DataBiosphere/data-explorer/pull/180).

It would be nice if elasticsearch-dsl could handle this automatically. Maybe TermsFacet could have an option convert_to_nested_if_applicable.

**Comment by HonzaKral** [comment]
I am sorry but as of now we try to stay away from an automatic functionality like this. I understand that this makes it difficult for code like yours which is trying to stay (afaict) generic.

While we are not planning to implement this, maybe we can implement something similar piggybacking on `Index.resolve_field` and provide a method like `resolve_with_nested` that would return a field along with a list of nested prefixes it belongs to. Using that to construct `Facet` instances should then be much easier.

Would that work for you?

**Comment by melissachang** [comment]
Yes, that would help, assuming `resolve_with_nested()` is fast. Then we wouldn't have to [construct nested paths ourselves](https://github.com/DataBiosphere/data-explorer/blob/504ea3de6243a22a44a96a811d0e29053dea6f8b/api/data_explorer/util/elasticsearch_util.py#L196-L250).

As a side note, as a user, it's unfortunate we have to jump through so many hoops (not just this issue) to get nested things to work. AFAICT, most people want [nested functionality](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html). It's hard for me to think of a use case where someone _wouldn't_ want nested functionality.

**Comment by HonzaKral** [comment]
> As a side note, as a user, it's unfortunate we have to jump through so many hoops (not just this issue) to get nested things to work. AFAICT, most people want nested functionality. It's hard for me to think of a use case where someone wouldn't want nested functionality.

That is interesting as `nested` usage is relatively rare in the wild, can I ask what is your primary use case for nested, can you share some sample documents? Thank you!

**Comment by melissachang** [comment]
We are using Elasticsearch to perform faceted search over a BigQuery table. For [BigQuery repeated records](https://cloud.google.com/bigquery/docs/nested-repeated), we use Elasticsearch nested object so faceted search works correctly.

For this BigQuery table:
![bigquery](https://user-images.githubusercontent.com/10929390/48581495-d6d92300-e8d6-11e8-8fb6-a062461470bc.png)

An Elasticsearch document might look like:
```
"_id" : "123",
"_source" : {
  "project_id.dataset_id.table_name.addresses" : [
    {
      "project_id.dataset_id.table_name.addresses.zip" : "11111",
      "project_id.dataset_id.table_name.addresses.city" : "New York"
    }
    {
      "project_id.dataset_id.table_name.addresses.zip" : "22222",
      "project_id.dataset_id.table_name.addresses.city" : "Washington, D.C."
    }
  ],
}
```
where `project_id.dataset_id.table_name.addresses` is nested.

I have the opposite question. Can you give me an example of a nested object where you do **not** want field to be nested? For example, in the [nested example](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html):
```
PUT my_index/_doc/1
{
  "group" : "fans",
  "user" : [ 
    {
      "first" : "John",
      "last" :  "Smith"
    },
    {
      "first" : "Alice",
      "last" :  "White"
    }
  ]
}
```
`user` should be `nested` instead of `object`. Can you give an example where the field should be `object` instead of `nested`?

**Comment by HonzaKral** [comment]
You only need `nested` if you have a list of objects and you need to query across more than one attribute at the same time (`zip` is `22222` and `city` is `Washington, D.C.`) otherwise plain `object` will work and there is no reason for `nested` queries and/or facets.

For your data model `nested` seems the right choice indeed as it allows for most flexibility. If, however, you;d only be interested in a single field at a time in facets/filters (give me all records that have a `Washington, D.C.` address) then you can get by without using `nested` as the flattened data (0) will still give you that functionality.

0 - https://www.elastic.co/guide/en/elasticsearch/reference/6.5/nested.html#_how_arrays_of_objects_are_flattened

**Comment by melissachang** [comment]
But Elasticsearch is all about querying. Are you saying most people don't need `nested` because they don't need to query across multiple fields? What if one user wants to query across multiple fields one day, how can someone say that will never happen? Like if my index is for logs analysis, I want to let users to be able to do whatever queries they want.

The following would make much more sense to me:
1. If there is a nested object, the type is `nested` instead of `object`.
1. TermFacets automatically become NestedFacets when appropriate

In fact, there should not even be a concept of `nested` vs `object`, `NestedFacet` vs `TermFacet`. All nested things should work out of the box.

(I'm guessing these concepts only exist because the first version of Elasticsearch didn't support them; they were added in later.)

Is there ever a scenario where you **wouldn't** want 1 and 2 from above?

**Comment by HonzaKral** [comment]
Actually most people don't have list of objects very often. Of course it is impossible to generalize and `nested` is an important feature but it definitely isn't major.

As I said `nested` is definitely the right choice for your use case and you made the right call, I was only offering the alternative in case it would work for you as it provides more simplicity and even some performance boost. I apologize if I misled you.

`object` is preferred if there is no array of objects and it is only just one (a Person only has one name with field `last`, `first`, and `middle`), there `nested` would provide no benefit and standard Facets would work just fine even when querying across multiple fields.

**Comment by melissachang** [comment]
Ok, that makes sense that list of objects is rare.

But if your index has lists of objects, is there a ever a scenario where you **wouldn't** want 1 and 2 from above? I argue that the answer is `no`; that 1 and 2 should be the default; so users with lists of objects don't have to deal with the pain that I'm going through.

This may not be a high priority feature request given that lists of objects is rare, but the feature request makes sense to me.

**Comment by HonzaKral** [comment]
Thank you for the feedback, I implemented the helper on `Index` in 845a2d6.

> But if your index has lists of objects, is there a ever a scenario where you wouldn't want 1 and 2 from above? I argue that the answer is no; that 1 and 2 should be the default; so users with lists of objects don't have to deal with the pain that I'm going through.

You are absolutely correct.

Right now there is no direct link between `FacetedSearch` and either `Document` or `Index` classes so the link would be difficult, if you have a reference to a `Document` what you can do is create a custom `FacetedSearch` subclass:

```python
class NestedSearch(FacetedSearch):
    def __init__(self, *args, _**kwarg):
        facets = {}
        index = self.doc_types[0]._index
        for f, facet in self.facets.items():
            nested, field = index.resolve_nested(facet._param['field'])
            for n in reversed(nested):
                facet = NestedFacet(n, facet)
            facets[f] = facet
        self.facets = facets
        super().__init__(*args, **kwargs)
```

which should be enough.

Hope this helps!
